src/libmeasurement_kit/ooni/templates.cpp

```diff
@@ -93,38 +93,51 @@ void http_request(Var<Entry> entry, Settings settings, http::Headers headers,
     http::request(
         settings, headers, body,
         [=](Error error, Var<http::Response> response) {
-            Entry rr;
-            /*
-             * Note: `probe_ip` comes from an external service, hence
-             * we MUST call `represent_string` _after_ `redact()`.
-             */
-            for (auto pair: headers) {
-                rr["request"]["headers"][pair.first] =
-                    represent_string(redact(pair.second));
-            }
-            rr["request"]["body"] = represent_string(redact(body));
-            rr["request"]["url"] = settings.at("http/url").c_str();
-            rr["request"]["method"] = settings.at("http/method").c_str();
 
-            // XXX we should probably update OONI data format to remove this.
-            rr["method"] = settings.at("http/method").c_str();
+            auto dump = [&](Var<http::Response> response) {
+                Entry rr;
+
+                if (!!error) {
+                    rr["failure"] = error.as_ooni_error();
+                } else {
+                    rr["failure"] = nullptr;
+                }
+
+                if (!!response) {
+                    /*
+                     * Note: `probe_ip` comes from an external service, hence
+                     * we MUST call `represent_string` _after_ `redact()`.
+                     */
+                    for (auto pair : response->headers) {
+                        rr["response"]["headers"][pair.first] =
+                            represent_string(redact(pair.second));
+                    }
+                    rr["response"]["body"] =
+                        represent_string(redact(response->body));
+                    rr["response"]["response_line"] =
+                        represent_string(redact(response->response_line));
+                    rr["response"]["code"] = response->status_code;
+
+                    auto request = response->request;
+                    for (auto pair : request->headers) {
+                        rr["request"]["headers"][pair.first] =
+                            represent_string(redact(pair.second));
+                    }
+                    rr["request"]["body"] =
+                        represent_string(redact(request->body));
+                    rr["request"]["url"] = request->url.str();
+                    rr["request"]["method"] = request->method;
+                }
+                return rr;
+            };
 
-            if (!error) {
-                for (auto pair: response->headers) {
-                    rr["response"]["headers"][pair.first] =
-                        represent_string(redact(pair.second));
+            if (!!response) {
+                for (Var<http::Response> x = response; !!x; x = x->previous) {
+                    (*entry)["requests"].push_back(dump(x));
```

Comment:
From what I can tell each session pair (request, response) will always have the same request sent. So for example if I have the following chain of redirects:
```
-> GET http://example.com/one
<- Location: http://example.com/two
-> GET http://example.com/two
<- Location: http://other.example.com/three
-> GET http://other.example.com/three
<- Body: other.example.com/three
```

What I will read in the report will actually be:

```json
{
   "requests": [
      {
         "request": {"url": "http://example.com/one"},
         "response": {"body": "other.example.com/three"},
     },
     {
       "request": {"url": "http://example.com/one"},
       "response": {"location": "http://example.com/three"},
     },
     {
       "request": {"url": "http://example.com/one"},
       "response": {"location": "http://example.com/two"},
     },
  ]
}
```
Note: am I getting the ordering correct? (This is the ordering we use in ooniprobe, item `[0]` is the last response)

What I would have expected instead would have been:

```json
{
   "requests": [
      {
         "request": {"url": "http://example.com/three"},
         "response": {"body": "other.example.com/three"},
     },
     {
       "request": {"url": "http://example.com/two"},
       "response": {"location": "http://example.com/three"},
     },
     {
       "request": {"url": "http://example.com/one"},
       "response": {"location": "http://example.com/two"},
     },
  ]
}
```